qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py

```diff
@@ -102,7 +84,13 @@ def run(self, dag):
                 new_circs.append(decomposer._decompose(operator))
             if new_circs:
                 new_circ = min(new_circs, key=len)
-                if len(run) > len(new_circ) or (single_u3 and new_circ.data[0][0].name != "u3"):
+                if all(g.name in self.basis for g in run) and len(run) < len(new_circ):
```

Comment:
Do you have an idea on what the runtime cost of this additional check is? If it's noticeable I'd rather not add this to just raise a warning because this will be called quite a lot during a transpile call.

Reply:
I don't have data, but I imagine it's negligible relative to the nonnegotiable cost of constructing `new_circ`: `self.basis` tends not to be super large, and `decomposer` has to read all of `run` anyhow. I'll do a small amount of with/without benchmarking.